pecos/core/utils/clustering.hpp

```diff
@@ -155,14 +203,46 @@ struct Tree {
     }
 
     template<typename MAT>
-    void partition_kmeans(size_t nid, size_t depth, const MAT& feat_mat, rng_t& rng, size_t max_iter=10, int threads=1, int thread_id=0) {
-        Node& root = root_of(nid);
-        Node& left = left_of(nid);
-        Node& right = right_of(nid);
+    void partition_kmeans(size_t nid, size_t depth, const MAT& feat_mat, rng_t& rng, size_t max_iter=10, int threads=1, int thread_id=0, float cur_sample_rate=1.0) {
+        // copy nodes rather than reference for sampling
+        Node root = root_of(nid);
+        Node left = left_of(nid);
+        Node right = right_of(nid);
+
+        // modify nodes' start and end based on cur_sample_rate 
+        size_t n_sp_elements = 0;
+        if(cur_sample_rate < 1.0) {
+            rng.shuffle(elements.begin() + root.start, elements.begin() + root.end);
+            n_sp_elements = int(cur_sample_rate * (root.end - root.start));
+            root.set(root.start, root.start + n_sp_elements);
+        }
         partition_elements(root, left, right);
 
         f32_sdvec_t& cur_center = center1[thread_id];
 
+        auto do_assignment = [&] (Node& root) {
```

Comment:
I noticed the `do_assignment()` function in `partition_kmeans` seems to be the same as the one in `partition_skmeans`? If so, can we just write one code, and reuse?